show collection chart for floor price and top sold nft

devVenus1202 opened the pull request:
**Thank you for your contribution** to the KodaDot NFT gallery.
Let's do a quick check before the merge.
Your contribution will be an eternal part of the community codebase.

### PR type
- [ ] Bugfix
- [x] Feature
- [ ] Refactoring

### Before submitting this PR, please make sure:
- [x] Code builds clean without any errors or warnings
- [x] I've merged recent default branch -- **main** and I've no conflicts
- [x] I've didn't break any original functionality
- [x] I've posted screenshot of demonstrated change in this PR

### Optional
- [x] I've tested it on mobile and everything works
- [ ] I found edge cases

### What's new? (may be part of changelog)
- This PR resolves  https://github.com/kodadot/nft-gallery/issues/909 - Bring history chart to collection

### If issue had bounty label ($,$$,$$$)
- Fill up your KSM address: https://kodadot.xyz/transfer/?target=HPQZ5yS3oKSwQwpiReGuo2Ug7qm9RR64bZXMfNbh8rD8Mwa 

### Community participation
- [x] Are you at KodaDot Discord? If not yet, [let's join us - https://discord.gg/35hzy2dXXh](https://discord.gg/35hzy2dXXh)

### Screenshot
- [x] Whenever my fix has changed **something** on UI, a screenshot for others, is more than welcome

devVenus1202 commented:
![image](https://user-images.githubusercontent.com/47501016/140615574-416c236f-c59b-4f2b-a0aa-97dbffe5aca4.png)

yangwao commented:
I know you are a first-time contributor, but can you please edit your PR description and link existing PR which is this for?

yangwao commented:
https://deploy-preview-1052--nftkodadot.netlify.app/rmrk/collection/2644199cf3652aaa78-KK01
Probably you should add some legend, how ppl would know what is for that lines?
![image](https://user-images.githubusercontent.com/5887929/140623335-ea2d77ed-4c8a-4587-9738-f753b90400cb.png)

https://deploy-preview-1052--nftkodadot.netlify.app/rmrk/collection/66241b171becfd500e-KU
![image](https://user-images.githubusercontent.com/5887929/140623365-ed1e3bfb-71eb-469a-b434-cc9754caf920.png)

vikiival commented:
Will do proper review in the morning:


- missing lints
- inconsistent code (lot of let, snake_case)
- lot of the code is already in the utils
- minor side effects

devVenus1202 commented:
@yangwao  I added legends as Floor Price and  Sold NFT Price.

yangwao commented:
> I know you are a first-time contributor, but can you please edit your PR description and link existing PR which is this for?

@devVenus1202 check this

roiLeo commented:
I think we should be using tabs to display activity chart instead of having to scroll to bottom.

![Screenshot 2021-11-08 at 09-49-46 Kusama Kings](https://user-images.githubusercontent.com/9987732/140711867-efd3848e-4302-42ec-aba2-6f9e249f2701.png)
:

devVenus1202 commented:
@yangwao I updated PR description again.

vikiival commented on src/components/rmrk/Gallery/CollectionItem.vue:
```diff
@@ -216,9 +221,40 @@ export default class CollectionItem extends Vue {
+    let topSoldPriceData : any[] = []
+
+    over_time.map((time: string) => {
+      let listEventsBeforeTime = this.listEventsBeforeTime(priceEvents, time).flat()
```
this should go into the utils

vikiival commented on src/components/rmrk/Gallery/CollectionItem.vue:
```diff
@@ -216,9 +221,40 @@ export default class CollectionItem extends Vue {
+
+    let buyEvents = events.map(this.onlyBuyEvents)?.flat().sort(sortByTimeStamp)
+    buyEvents?.map((e:any) => {
+      topSoldPriceData.push([new Date(e.timestamp), Number(e.meta) / 1000000000000])
```
incorectly used constant (10 ** 12)

vikiival commented on src/components/rmrk/Gallery/CollectionItem.vue:
```diff
@@ -244,5 +280,31 @@ export default class CollectionItem extends Vue {
+
+  protected onlyEvents(nft:NFT){
+    return nft.events.map((e:Interaction) => {
+      return {...e, name: nft.name}
```
this should go into the utils

vikiival commented on src/components/rmrk/Gallery/CollectionItem.vue:
```diff
@@ -244,5 +280,31 @@ export default class CollectionItem extends Vue {
+      return {...e, name: nft.name}
+    })
+  }
+  protected priceEvents(nftEvents:Interaction[]){
```
What should this do ?

vikiival commented on src/components/rmrk/Gallery/CollectionItem.vue:
```diff
@@ -244,5 +280,31 @@ export default class CollectionItem extends Vue {
+  protected priceEvents(nftEvents:Interaction[]){
+    return nftEvents.filter((event:Interaction) => event.interaction === 'LIST' || event.interaction === 'BUY' || event.interaction === 'SEND' || event.interaction === 'CONSUME')
+  }
+  protected onlyBuyEvents(nftEvents:any[]){
```
this is already in utils

devVenus1202 replied:
Can you point where save function in utils?
Thanks!

roiLeo replied:
@devVenus1202 search for 
```js 
function pairListBuyEvent
```

vikiival replied:
still unresolved

vikiival replied:
so the logic behind is that current daily floor is calculated as:

```ts
Math.min(...this.nfts.map(nft => Number(nft.price)).filter(price => price > 0))
```

but this won't work because current state automatically filters out burned.

For the floor per day: We need to check other events because as @devVenus1202 pointed out it changes the floor price.

But still I would omit the usage of three or `||` in the condition.

Third: Please write a test for this

vikiival commented on src/components/rmrk/Gallery/CollectionItem.vue:
```diff
@@ -216,9 +221,40 @@ export default class CollectionItem extends Vue {
+    let events : Interaction[][] = this.stats?.map(this.onlyEvents) || []
+    let priceEvents : Interaction[][] = events.map(this.priceEvents) || []
+
+    let over_time : string[] = priceEvents.flat().sort(sortByTimeStamp).map((e: Interaction) => e.timestamp)
```
⚠️ use camelCase

vikiival reviewed (changes requested):
Generally a lot of: 
-  `let`
- code that can be abstracted to the utils
- snake_case
- missing types
- badly used constant (10**12 works only for kusama)

devVenus1202 commented:
I will update the points

vikiival commented on src/components/rmrk/Gallery/CollectionItem.vue:
```diff
@@ -163,6 +167,14 @@ export default class CollectionItem extends Vue {
     return !isShareMode
   }
 
+  get chainProperties() {
```
please use `ChainMixin`

vikiival commented on src/components/rmrk/utils.ts:
```diff
@@ -294,11 +295,29 @@ export const resolveMedia = (mimeType?: string): MediaType => {
+  return [new Date(time), floorPrice]
+}
+
+export const soldNFTPrice = (decimals : number) => (e : Interaction) => [new Date(e.timestamp), Number(e.meta) / 10 ** decimals]
```
pls use linter

vikiival reviewed (changes requested):
Better, still some minor issues

yangwao commented:
> I think we should be using tabs to display activity chart instead of having to scroll to bottom.
> 
> ![Screenshot 2021-11-08 at 09-49-46 Kusama Kings](https://user-images.githubusercontent.com/9987732/140711867-efd3848e-4302-42ec-aba2-6f9e249f2701.png) :

Yes, was thinking about this with #1037 to bit sort stuff.

@devVenus1202 can you please add tabs there?
I'll raise bounty on this issue, but please keep your code quality high! :)

vikiival reviewed (commented):
Resolved issue with floor:

so the logic behind is that current daily floor is calculated as:

Math.min(...this.nfts.map(nft => Number(nft.price)).filter(price => price > 0))
but this won't work because current state automatically filters out burned.

For the floor per day: We need to check other events because as @devVenus1202 pointed out it changes the floor price.

But still I would omit the usage of three or || in the condition.

Third: Please write a test for this

devVenus1202 commented:
Hi @vikiival 
I fixed the issues and wrote a test for a floor price - floorPrice.spec.ts.
Please check it for me.

vikiival reviewed (changes requested):
Forgot to submit

devVenus1202 commented:
@yangwao I will make tabs for activity chart.
Can you point where I can make tabs in collection page or give me the url for reference?

devVenus1202 commented on tests/unit/floorPrice.spec.ts:
```diff
@@ -0,0 +1,35 @@
+
+    const now = new Date()
+
+    const priceEvents: Interaction[] = nfts
```
Okay, I will reuse the functions from the utils.

roiLeo commented:
> @yangwao I will make tabs for activity chart. Can you point where I can make tabs in collection page or give me the url for reference?

We're using Buefy, you can use [Tabs](https://buefy.org/documentation/tabs/#position) component

devVenus1202 commented:
Hi @vikiival 
Please check updated PR again.
Additionally, I added tabs for collection activity.

yangwao commented:
Let's wait on merging https://github.com/kodadot/nft-gallery/pull/1079
meanwhile please resolve conflicts @devVenus1202

devVenus1202 commented:
Okay, I will resolve.

devVenus1202 commented:
@yangwao I resolved the conflicts.

yangwao commented:
Hey, @devVenus1202 plus @KurtWarsonMillenium I have serious doubts you are far from the senior full-stack developer. Using two handles in committing and the other for PR communication is super confusing.

I see this at https://deploy-preview-1052--nftkodadot.netlify.app/rmrk/collection/4a5668af20c28c074f-KSMDUCKZ
![image](https://user-images.githubusercontent.com/5887929/141177304-d8d89191-59d5-4c7c-b0ef-5b137c16d099.png)

devVenus1202 commented:
Hey, @yangwao. @KurtWarsonMillenium is my client's account.
I switched git account from this but commits are handled by @KurtWarsonMillenium 
It is really weird and sorry for confusing to you.
If you give me more opportunity, I will commit with only @devVenus1202 - my account in other PRs.

And I removed duplicated collectionActivity component.

yangwao commented:
Hey, there were still some issues in code quality and I want to also close on it. So fast-tracking it for now.
It's working but the quality is not best, please mind that in next time making PR, please. 
Payout sent 
https://kusama.subscan.io/extrinsic/0xeec2555792387107de91e11df2e8dc6f2237da0828e4e8c356137841c7455fec
